src/Models/Services/TelemetryService/TelemetryService.ts

```diff
@@ -1,21 +1,106 @@
+import { Md5 } from 'ts-md5';
+import { LOCAL_STORAGE_KEYS } from '../../Constants/Constants';
+import { CUSTOM_PROPERTY_NAMES } from '../../Constants/TelemetryConstants';
 import { getDebugLogger } from '../Utils';
-import { TelemetryItem } from './Telemetry';
+import {
+    TelemetryEvent,
+    TelemetryException,
+    TelemetryItem,
+    TelemetryMetric,
+    TelemetryRequest,
+    TelemetryTrace
+} from './Telemetry';
 
-const debugLogging = false;
+const debugLogging =
+    localStorage.getItem(
+        LOCAL_STORAGE_KEYS.FeatureFlags.Telemetry.debugLogging
+    ) === 'true' || false;
 const logDebugConsole = getDebugLogger('TelemetryService', debugLogging);
 
 class TelemetryService {
-    static telemetryCallback: (telemetry: TelemetryItem) => void;
+    static telemetryCallback: (telemetry: TelemetryItem) => Promise<void>;
+    private static adtInstanceHash: string; // the hashed value of the adt instance URL
+    private static storageContainerHash: string; // the hashed value of the storage container url
+    private static sceneHash: string; // the hash of the scene id
 
     // Attach telemetry processing callback from consuming application
     static registerTelemetryCallback(
-        telemetryCallback: (telemetry: TelemetryItem) => void
+        telemetryCallback: (telemetry: TelemetryItem) => Promise<void>
     ) {
         TelemetryService.telemetryCallback = telemetryCallback;
     }
 
+    // #region public telemetry emitters
+
+    static sendEvent(telemetry: TelemetryEvent) {
+        this.sendTelemetry(telemetry);
+    }
+
+    static sendException(telemetry: TelemetryException) {
+        this.sendTelemetry(telemetry);
+    }
+
+    static sendMetric(telemetry: TelemetryMetric) {
+        this.sendTelemetry(telemetry);
+    }
+
+    static sendRequest(telemetry: TelemetryRequest) {
+        this.sendTelemetry(telemetry);
+    }
+
+    static sendTrace(telemetry: TelemetryTrace) {
+        this.sendTelemetry(telemetry);
+    }
+
+    // #endregion
+
+    // #region public setters
+
+    /**
+     * store the current adt instance
+     * @param adtInstanceUrl
+     */
+    static setAdtInstance(adtInstanceUrl: string): void {
+        this.adtInstanceHash = Md5.hashStr(adtInstanceUrl);
```

Comment:
just wondering where do we keep the mapping for hash to url? I mean after keeping the urls as hash in the telemetry how to we backtrack to the urls for debugging purposes?

Reply:
We don't that's very intentional. We don't want that private data in clear text anywhere in our telemetry. We won't be able to reverse the hash without the original value to hash again. The point is to provide a unique and consistent id without being able to tie it to a particular customer.